Userland/Libraries/LibAudio/FlacLoader.cpp

```diff
@@ -0,0 +1,899 @@
+        // if (left[i] != 0)
+            // dbgln("Frame data {:8} : {:8} -> {:7.4f} {:7.4f}", left[i], right[i], frame.left, frame.right);
+        m_current_frame_data.unchecked_append(frame);
+    }
+
+#undef CHECK_OK
+#undef CHECK_ERROR_STRING
+}
+
+u32 FlacLoaderPlugin::convert_sample_count_code(u8 sample_count_code)
+{
+    // single codes
+    switch (sample_count_code) {
+    case 0:
+        m_error_string = "Reserved block size";
+        return 0;
+    case 1:
+        return 192;
+    case 6:
+        return FLAC_BLOCKSIZE_AT_END_OF_HEADER_8;
+    case 7:
+        return FLAC_BLOCKSIZE_AT_END_OF_HEADER_16;
+    }
+    if (sample_count_code >= 2 && sample_count_code <= 5) {
+        return 576 * pow(2, (sample_count_code - 2));
+    }
+    return 256 * pow(2, (sample_count_code - 8));
+}
+
+u32 FlacLoaderPlugin::convert_sample_rate_code(u8 sample_rate_code)
+{
+    switch (sample_rate_code) {
+    case 0:
+        return m_sample_rate;
+    case 1:
+        return 88200;
+    case 2:
+        return 176400;
+    case 3:
+        return 192000;
+    case 4:
+        return 8000;
+    case 5:
+        return 16000;
+    case 6:
+        return 22050;
+    case 7:
+        return 24000;
+    case 8:
+        return 32000;
+    case 9:
+        return 44100;
+    case 10:
+        return 48000;
+    case 11:
+        return 96000;
+    case 12:
+        return FLAC_SAMPLERATE_AT_END_OF_HEADER_8;
+    case 13:
+        return FLAC_SAMPLERATE_AT_END_OF_HEADER_16;
+    case 14:
+        return FLAC_SAMPLERATE_AT_END_OF_HEADER_16X10;
+    default:
+        m_error_string = "Invalid sample rate code";
+        return 0;
+    }
+}
+
+PcmSampleFormat FlacLoaderPlugin::convert_bit_depth_code(u8 bit_depth_code)
+{
+    switch (bit_depth_code) {
+    case 0:
+        return m_sample_format;
+    case 1:
+        return PcmSampleFormat::Uint8;
+    case 4:
+        return PcmSampleFormat::Int16;
+    case 6:
+        return PcmSampleFormat::Int24;
+    case 3:
+    case 7:
+        m_error_string = "Reserved sample size";
+        return PcmSampleFormat::Float64;
+    default:
+        m_error_string = String::formatted("Unsupported sample size {}", bit_depth_code);
+        return PcmSampleFormat::Float64;
+    }
+}
+
+u8 frame_channel_type_to_channel_count(FlacFrameChannelType channel_type)
+{
+    if (channel_type <= 7)
+        return channel_type + 1;
+    return 2;
+}
+
+FlacSubframeHeader FlacLoaderPlugin::next_subframe_header(InputBitStream& bit_stream, u8 channel_index)
+{
+    u8 bits_per_sample = pcm_bits_per_sample(m_current_frame->bit_depth);
+
+    // For inter-channel correlation, the side channel needs an extra bit for its samples
+    switch (m_current_frame->channels) {
+    case LeftSideStereo:
+    case MidSideStereo:
+        if (channel_index == 1) {
+            ++bits_per_sample;
+        }
+        // COMBAK: handle higher bits per sample for side channels, also in the decoders below
```

Comment:
"COMBAK"? :thinking:

Reply:
Comeback note for myself, also this is resolved so I'll go remove it.